### PR TITLE
Loki: query editor using Monaco

### PR DESCRIPTION
**Pull request by matyax**
**What this PR does / why we need it**:

This PR resumes the work from this [hackathon PR](https://github.com/grafana/grafana/pull/46311), to migrate the Loki query editor to Monaco.

To achieve that, the original branch has been updated, and some refactors, fixes, improvements, and missing pieces were added.

**Which issue(s) this PR fixes**:

Part of https://github.com/grafana/grafana/issues/44985

A follow-up PR will add:
- Tests for the components.
- `extractedLabelKeys` implementation.
- Improve "loading" display (when lazy loading the component).
- Add support for the `IN_AGGREGATION` situation.

**Special notes for your reviewer**:

It's currently under a feature flag, so you'll need to enable it to test it.

### CODE REVIEW

File: public/app/plugins/datasource/loki/LanguageProvider.ts
Comment (matyax) [inline public/app/plugins/datasource/loki/LanguageProvider.ts:527]: @gabor Could you help me understand these `FIXME` here?

Comment (gabor) [reply public/app/plugins/datasource/loki/LanguageProvider.ts:527]: @matyax at that point in the code, i need to create a `DataQueryRequest` to fetch some data from the database. so i have set up all the "unnecessary" attributes to some default-values to make typescript happy, but i added `FIXME` because i'm not sure if those values are really correct. for example, we should probably take the timezone-value from what is chosen in the grafana time-range-picker.

Comment (gabor) [reply public/app/plugins/datasource/loki/LanguageProvider.ts:527]: but, that code might not be necessary anymore.. you see, the idea there was to fetch a couple lines of logs from Loki, "understand" those log-lines, and based on that offer smart autocomplete options about how to finish the logql-query.

but, some time ago the loki-query-builder got the query-hint functionality, that does basically the same.
for example, in the Loki query builder (assuming you are running `devenv` Loki), if you choose the labelName:`place`, labelValue:`luna` combination, it will offer you to add the `logfmt` parser. it does this by fetching a couple lines of logs and "understanding" them. this is the same as the one done by the monaco-loki-query-field, so i think we should be able to somehow "connect" to that functionality and use it too.

Comment (matyax) [reply public/app/plugins/datasource/loki/LanguageProvider.ts:527]: Thank you for the explanation. I'll check if I can use that here.

File: public/app/plugins/datasource/loki/LanguageProvider.ts
Comment (gabor) [reply public/app/plugins/datasource/loki/LanguageProvider.ts]: @matyax nice work! (i know this has changed since this comment, i checked the actual version)

so, `extractedLabelKeys`... i assume the question is: from the data-samples, i'd like to extract all the label-keys, correct?

you will need to go through the dataframe-array returned by `datasource.getDataSamples`. in loki dataframes, there is a field (it's a "column" of data), named "labels", and it's contents are javascript-objects that contain the key-value pairs for the given row.

so something like:
```ts
const allKeysDuplicated = series.forEach(frame => {
   const labelsField = frame.fields.find(f => f.name === 'labels');
   const objs = labelsField.values.toArray();
   const keys = objs.map(obj => Object.keys(obj));
   return keys;
});
const allKeys = Array.from(new Set(allKeysDuplicated));
```

i wrote this blindly, i hope it explains what i mean, if not just ping me :)

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.test.ts
Comment (matyax) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.test.ts:79]: @gabor @ivanahuckova Gonna need some help here.

The current code in `situation.ts` expects the following path for `sum(`: `[ERROR_NODE_NAME, 'VectorAggregationExpr']`, and the parser is returning:

```
[
    'ERROR_NODE_NAME',
    'LiteralExpr',
    'MetricExpr',
    'VectorAggregationExpr',
    'MetricExpr',
    'Expr',
    'LogQL'
]
```

I'm not sure if I need to update the path with this other one, or if there is a bug in the code.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.test.ts
Comment (matyax) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.test.ts:116]: @gabor @ivanahuckova I had to remove the test case for single quotes (`{one='val\'1',^}`) from the test, because currently `lezer-logql` process the query as errors:

imagen

Unline double quotes or ticks:

imagen

Not sure if this is expected or not. In the hackathon PR there was a test case for single quotes.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.test.ts
Comment (gabor) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.test.ts:79]: @matyax in general, since the hackathon-code is quite old, if lezer-logql produces a different tree currently, i would just update the "hackathon-code".

but please make sure to always have tests in `situation.test.ts`, the largest reason why there is a sepratate `situation.ts` and a `completions.ts` is that `situation.ts` is very easy/efficient to unit-test.

also, it made sense from a logical side (one deals with "find out what is the situation we are in", other deals with "what should we offer in this situation")

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.test.ts
Comment (gabor) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.test.ts:116]: @matyax , that may be a mistake in the hackathon-code. i have just tried to run a query like this in Loki `{place='on\'e'}`, and it said 
"parse error at line 1, col 8: invalid char literal"

so, maybe `\'` does not exist :) . pinging @ivanahuckova for a definitive answer.

**Comment by ivanahuckova** [comment]
I am wondering if there is a way how this PR could be split into smaller PRs?

**Comment by matyax** [comment]
I don't think so, for 2 reasons. 1-, it's coming from an already "big" PR, and 2- the whole set of changes is required for the feature.

If you think it will help you, I can write a summary of the changes, and what you can find on each file.

File: public/app/plugins/datasource/loki/components/monaco-query-field/getOverrideServices.ts
Comment (matyax) [inline public/app/plugins/datasource/loki/components/monaco-query-field/getOverrideServices.ts:1]: This file is exactly the same as https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/components/monaco-query-field/getOverrideServices.ts. I considered removing it and importing it from there, or something similar.

**Comment by matyax** [comment]
The main files to review are the following:

- MonacoQueryField.tsx
- Everything inside monaco-completion-provider (4 files, the rest are tests)
- syntax.ts

The rest are tests, a feature flag, some wrapper components, and minor changes.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts
Comment (matyax) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts:1]: This is a data provider for completions that sits between the editor code and the `languageProvider` instance.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts
Comment (matyax) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts:1]: These are the functions that, based in a given situation, select which completion options to return.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts
Comment (matyax) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts:1]: Here is the code to resolve from an input value to an autocompletion "situation". Each situation will require a different autocompletion set of strings.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/index.ts
Comment (matyax) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/index.ts:1]: This is Monaco specific-code, which glues the editor instance with the code in `completion.ts`.

File: public/app/plugins/datasource/loki/datasource.ts
Comment (gabor) [inline public/app/plugins/datasource/loki/datasource.ts:291]: do you want to use `this.getTimeRange()` at https://github.com/grafana/grafana/blob/matyax/monaco-loki-from-hackathon/public/app/plugins/datasource/loki/datasource.ts#L259 too?

File: pkg/build/cmd/rpm.go
Comment (gabor) [inline pkg/build/cmd/rpm.go:294]: so, we have no idea why this shows up in the diff 😄 . the version "after" is what is currently in the main-branch already.

**Review by gabor — APPROVED** [review]
LGTM! nice work 🎉

**Comment by matyax** [comment]
Thank you @gabor ! We'd appreciate another pair of eyes in this PR before merging @grafana/observability-logs . No rush, of course, since this is an XL PR.

File: public/app/plugins/datasource/loki/components/LokiQueryField.tsx
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/LokiQueryField.tsx:216]: Nit/personal preference I'd destructure props and just do  `runQueryOnBlur={app !== CoreApp.Explore}` everywhere.

File: pkg/build/cmd/rpm.go
Comment (svennergr) [reply pkg/build/cmd/rpm.go:294]: I guess this could be removed if we/ @matyax rebases the branch on main?

Comment (matyax) [reply pkg/build/cmd/rpm.go:294]: Probably, but this is a feature branch from an old feature branch with multiple main merges. Rebasing would be catastrophic.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts:20]: I don't think this history works anymore. Not related to this PR, but is always returns just empty array.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts:1]: This is probably not related import.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts:61]:
```suggestion
```

`$__rate_interval` is not supported for Loki.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts:195]: We shouldn't offer line filters if we are after first pipe.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts:42]: We should probably change "something" to something else :D Maybe just `|= ""` is enough.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts:89]: Why is unwrap `LINE_FILTER`. It is not line filter. It is kind of category of its own. So I would either create new category or create `PIPE_OPERATIONS` and add there all operations after | (besides parsers). 🤔

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts:6]: here instead of hardcoded values, we should use imported ids in the same way as we do in prometheus https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/components/monaco-query-field/monaco-completion-provider/situation.ts#L29

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts:128]:
```suggestion
const ERROR_NODE_NAME: NodeTypeId = 0; // this is used as error-id
```

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts:132]: Again, these hardcoded values should be replaced with imported ids from loki lezer.

**Review by ivanahuckova — COMMENTED** [review]
I did the first round of review and left some feedback. This is going to be great improvement!

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts:42]: Maybe even adding there documentation that it is line contains

File: pkg/services/featuremgmt/registry.go
Comment (matyax) [reply pkg/services/featuremgmt/registry.go:105]: This one is automagically generated. There's also a mixture of styles, not sure if we should change it or not.

File: public/app/plugins/datasource/loki/components/LokiQueryField.tsx
Comment (matyax) [reply public/app/plugins/datasource/loki/components/LokiQueryField.tsx:216]: Sure.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts
Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts:20]: We have a `history` prop that we pass in `LokiQueryField.tsx`. Can please you confirm that it will always be empty? And maybe add some context to better understand why it will not work, because maybe we should remove it altogether?

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts
Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts:1]: Yes, removed.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts
Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts:42]: Good catch.

> Maybe even adding there documentation that it is line contains

Not sure I understand.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts
Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts:87]: Could you please give me more context to understand this? Is there any reference I can use for this? Thanks!

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts
Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts]: Removed!

File: public/app/plugins/datasource/loki/syntax.ts
Comment (matyax) [reply public/app/plugins/datasource/loki/syntax.ts:178]: Removed.

**Comment by matyax** [comment]
Thanks for the feedback! There are still some things to adjust, but I could use more info in a couple of comments.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts:20]: Sorry, your code is correct, but the history is somehow broken, because I never get history autocomplete suggestions and when I console.log it, it is always empty. So nothing related to this PR, just a note that we should look into history autocomplete part. I'll create separate issue for this.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts]: Some functions/options have this additional explanation:
image

So maybe for `|= ""` we could add explanation that it is 'line contains'. And `|~ ""` is regex match.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts
Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts:20]: Ohh gotcha. Yes, I understood that part, but had doubts on whether we should remove it or not. Thanks!

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts:63]: Here is the documentation https://grafana.com/docs/loki/latest/logql/log_queries and here is a miro board I once created for this project: https://miro.com/app/board/uXjVOF5z1u8=/. 

The issue here is that unwrap is pipe operation in metrics query and it doesn't belong to any category mentioned in log queries.Here are categories for log operations: 
image

[Here is the documentation for unwrap](https://grafana.com/docs/loki/latest/logql/metric_queries/#unwrapped-range-aggregations) and it called `unwrap expression`.

So either `unwrap` should have category on its own in the same way as you created category for `label_format` and `line_format`, which I think would be preferable. Or we could create `PIPE_OPERATIONS` group that would include `line_format`, `label_format` and `unwrap.`

File: pkg/services/featuremgmt/registry.go
Comment (ivanahuckova) [reply pkg/services/featuremgmt/registry.go:105]: I've seen that mostly it uses Sentence case, but this is definitely non-blocking feedback. 🙂 Just a thing I noticed.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts]: 🙈 this is on my poor comment again. Now, the line filters completions are not suggested after selector at all. They should be suggested after selector, but as soon as user writes first pipe `|`, the line filters completions should be not displayed anymore. You can probably revert https://github.com/grafana/grafana/pull/55391/commits/3d003ca4bcb792b440add04b883fafe637e4c4c9 and tackle this case in separate PR.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts
Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts]: Good catch. Refactored everything to use node identifiers.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts
Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts]: Done.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts
Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts:174]: Done.

**Review by ivanahuckova — APPROVED** [review]
I think we can merge as this is behind feature flag and then fix autocompleting:

- Line filters are not suggested anymore
- Unwrap has incorrect category 
- Functions are not offered inside aggregations

In separate PRs

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts
Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts]: Oh, okay. Thanks for clarifying.

Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts]: Reverted.

**Comment by matyax** [comment]
Sounds good to me. I updated the issue to keep track of the missing parts, let me know if I skipped anything.

**Pull request merged**
